src/envoy/tcp/tcp_cluster_rewrite/tcp_cluster_rewrite.cc

```diff
@@ -63,12 +63,25 @@ Network::FilterStatus TcpClusterRewriteFilter::onNewConnection() {
                    "tcp_cluster_rewrite: final tcp proxy cluster name {}",
                    read_callbacks_->connection(), final_cluster_name);
 
-    // The data is mutable to allow other filters to change it.
-    read_callbacks_->connection().streamInfo().filterState()->setData(
-        TcpProxy::PerConnectionCluster::key(),
-        std::make_unique<TcpProxy::PerConnectionCluster>(final_cluster_name),
-        StreamInfo::FilterState::StateType::Mutable);
+    try {
+      // The data is mutable to allow other filters to change it.
+      read_callbacks_->connection().streamInfo().filterState()->setData(
+          TcpProxy::PerConnectionCluster::key(),
+          std::make_unique<TcpProxy::PerConnectionCluster>(final_cluster_name),
+          StreamInfo::FilterState::StateType::Mutable,
+          StreamInfo::FilterState::LifeSpan::DownstreamConnection);
+    } catch (const EnvoyException& e) {
```

Comment:
@PiotrSikora I thought Envoy didn't use exceptions on the data path.  Is this the right way to handle or should the code this is calling be changed to not through an exception?

Nice find @yxue!

Reply:
Or is this the control path? :)

Reply:
I think this is the data path. If the code swallow the exception here, the behavior will be unexpected because the tcp proxy depends on the data in the filter state to do routing.

Reply:
It throws when it's called twice with conflicting declarations (questionable design choice). The only other use I found was `sni_cluster` filter, but it matches the declaration here, so in theory it should never throw. I think it's fine to just remove the try/catch block. The danger is mis-configuration with two filters trying to declare the same filter state key in different ways.

Reply:
Mis configuration two filters with the same key is definitely wrong. Adding the try/catch is for debugging. Currently the exception is thrown inside the setData and difficult for pinpointing the issue. 

Current stack trace https://gist.github.com/ayj/5c07b20d651bb6be48a08b653f0ff590#file-log_trace-L604-L605

Reply:
It does rethrow but that might be subtle to a reader.  Perhaps document that this is just to get a better error message and not for recovery?